mlflow/langchain/_langchain_autolog.py

```diff
@@ -0,0 +1,296 @@
+        # `config` is the second positional argument of runnable.invoke function
+        if len(args) >= 2:
+            config = args[1]
+            in_args = True
+        else:
+            config = kwargs.get("config", None)
+        if config is None:
+            callbacks = [mlflow_callback]
+            config = RunnableConfig(callbacks=callbacks)
+        else:
+            callbacks = config.get("callbacks") or []
+            callbacks.append(mlflow_callback)
+            config["callbacks"] = callbacks
+        if in_args:
+            args = (args[0], config) + args[2:]
+        else:
+            kwargs["config"] = config
+        return args, kwargs
+
+    if func_name == "__call__":
+        # `callbacks` is the third positional argument of chain.__call__ function
+        if len(args) >= 3:
+            callbacks = args[2] or []
+            callbacks.append(mlflow_callback)
+            args = args[:2] + (callbacks,) + args[3:]
+        else:
+            callbacks = kwargs.get("callbacks") or []
+            callbacks.append(mlflow_callback)
+            kwargs["callbacks"] = callbacks
+        return args, kwargs
+
+    if func_name == "get_relevant_documents":
+        callbacks = kwargs.get("callbacks") or []
+        callbacks.append(mlflow_callback)
+        kwargs["callbacks"] = callbacks
+        return args, kwargs
+
+
+def _runnable_with_retriever(model):
+    from langchain.schema import BaseRetriever
+
+    with contextlib.suppress(ImportError):
+        from langchain.schema.runnable import RunnableBranch, RunnableParallel, RunnableSequence
+        from langchain.schema.runnable.passthrough import RunnableAssign
+
+        if isinstance(model, RunnableBranch):
+            return any(_runnable_with_retriever(runnable) for _, runnable in model.branches)
+
+        if isinstance(model, RunnableParallel):
+            return any(_runnable_with_retriever(runnable) for runnable in model.steps.values())
+
+        if isinstance(model, RunnableSequence):
+            return any(_runnable_with_retriever(runnable) for runnable in model.steps)
+
+        if isinstance(model, RunnableAssign):
+            return _runnable_with_retriever(model.mapper)
+
+    return isinstance(model, BaseRetriever)
+
+
+def _chain_with_retriever(model):
+    with contextlib.suppress(ImportError):
+        from langchain.chains import RetrievalQA
+
+        return isinstance(model, RetrievalQA)
+    return False
+
+
+def patched_inference(func_name, original, self, *args, **kwargs):
+    """
+    A patched implementation of langchain models inference process which enables logging the
+    following parameters, metrics and artifacts:
+
+    - model
+    - metrics
+    - data
+
+    We patch either `invoke` or `__call__` function for different models
+    based on their usage.
+    """
+
+    import langchain
+
+    # import from langchain_community for test purpose
+    from langchain_community.callbacks import MlflowCallbackHandler
```

Comment:
`MlflowCallbackHandler` logs parameters, metrics, and artifacts?

Reply:
It logs metrics, no parameters as all the parameters are part of the logged model

Reply:
Got it, thanks for the explanation!